src/bio/smd/smd_device.c

```diff
@@ -20,293 +20,294 @@
+	int			rc;
+
+	D_ASSERT(!daos_handle_is_inval(smd_store.ss_dev_hdl));
+	D_ASSERT(!daos_handle_is_inval(smd_store.ss_tgt_hdl));
+
+	uuid_copy(key_dev.uuid, dev_id);
+	D_RWLOCK_WRLOCK(&smd_store.ss_rwlock);
+
+	/* Check if the target is already bound to a device */
+	d_iov_set(&key, &tgt_id, sizeof(tgt_id));
+	d_iov_set(&val, &bond_dev, sizeof(bond_dev));
+	rc = dbtree_fetch(smd_store.ss_tgt_hdl, BTR_PROBE_EQ,
+			  DAOS_INTENT_DEFAULT, &key, NULL, &val);
+	if (rc == 0) {
+		D_ERROR("Target %d is already bound to dev "DF_UUID"\n",
+			tgt_id, DP_UUID(&bond_dev.uuid));
+		rc = -DER_EXIST;
+		goto out;
+	} else if (rc != -DER_NONEXIST) {
+		D_ERROR("Get target %d failed. %d\n", tgt_id, rc);
+		goto out;
+	}
 
-	ndev_df = umem_off2ptr(&tins->ti_umm, rec->rec_off);
-	memcpy(val_iov->iov_buf, ndev_df, sizeof(*ndev_df));
-	return 0;
-}
+	/* Fetch device if it's already existing */
+	d_iov_set(&key, &key_dev, sizeof(key_dev));
+	d_iov_set(&val, &entry, sizeof(entry));
+	rc = dbtree_fetch(smd_store.ss_dev_hdl, BTR_PROBE_EQ,
+			  DAOS_INTENT_DEFAULT, &key, NULL, &val);
+	if (rc == 0) {
+		if (entry.sde_tgt_cnt >= SMD_MAX_TGT_CNT) {
+			D_ERROR("Dev "DF_UUID" is assigned to too many "
+				"targets (%d)\n", DP_UUID(&key_dev.uuid),
+				entry.sde_tgt_cnt);
+			rc = -DER_OVERFLOW;
+			goto out;
+		}
+		entry.sde_tgts[entry.sde_tgt_cnt] = tgt_id;
+		entry.sde_tgt_cnt += 1;
+	} else if (rc == -DER_NONEXIST) {
+		entry.sde_state = SMD_DEV_NORMAL;
+		entry.sde_tgt_cnt = 1;
+		entry.sde_tgts[0] = tgt_id;
+	} else {
+		D_ERROR("Fetch dev "DF_UUID" failed. %d\n",
+			DP_UUID(&key_dev.uuid), rc);
+		goto out;
+	}
 
-static int
-dtab_df_rec_update(struct btr_instance *tins, struct btr_record *rec,
-		   d_iov_t *key_iov, d_iov_t *val_iov)
-{
-	struct smd_nvme_dev_df		*ndev_df;
-	int				 rc;
+	/* Update device and target tables in same transaction */
+	rc = smd_tx_begin(&smd_store);
+	if (rc)
+		goto out;
 
-	rc = umem_tx_add(&tins->ti_umm, rec->rec_off,
-			 sizeof(struct smd_nvme_pool_df));
-	if (rc != 0)
-		return rc;
+	rc = dbtree_update(smd_store.ss_dev_hdl, &key, &val);
+	if (rc) {
+		D_ERROR("Update dev "DF_UUID" failed. %d\n",
+			DP_UUID(&key_dev.uuid), rc);
+		goto tx_end;
+	}
 
-	ndev_df = umem_off2ptr(&tins->ti_umm, rec->rec_off);
-	memcpy(ndev_df, val_iov->iov_buf, val_iov->iov_len);
-	return 0;
+	d_iov_set(&key, &tgt_id, sizeof(tgt_id));
+	d_iov_set(&val, &key_dev, sizeof(key_dev));
+	rc = dbtree_update(smd_store.ss_tgt_hdl, &key, &val);
+	if (rc) {
+		D_ERROR("Update target %d failed. %d\n", tgt_id, rc);
+		goto tx_end;
+	}
+tx_end:
+	rc = smd_tx_end(&smd_store, rc);
+out:
+	D_RWLOCK_UNLOCK(&smd_store.ss_rwlock);
+	return rc;
 }
 
-btr_ops_t dtab_ops = {
-	.to_hkey_size	= dtab_df_hkey_size,
-	.to_hkey_gen	= dtab_df_hkey_gen,
-	.to_rec_alloc	= dtab_df_rec_alloc,
-	.to_rec_free	= dtab_df_rec_free,
-	.to_rec_fetch	= dtab_df_rec_fetch,
-	.to_rec_update	= dtab_df_rec_update,
-};
-
 int
-smd_nvme_dtab_register()
+smd_dev_unassign(uuid_t dev_id, int tgt_id)
```

Comment:
if this is not currently supported, should be mentioned in header file? is this not implemented because we want to maintain the list of devs->tgts even after failure? will this get implemented?

Reply:
I was thinking this only needed when we support reintegration, then we probably need to remove the old dev<->tgt mappings.

Reply:
I think we  need to cleanup when target is removed actually, just to close the circuit.